pages/react-canvas-components.mdx

```diff
@@ -0,0 +1,256 @@
+    const context2d = canvasRef.current.getContext('2d');
+    setRenderingContext(context2d);
+  }, []);
+
+  return (
+    <SharingContext.Provider value={renderingContext}>
+      <canvas ref={canvasRef} />
+      {/* the hexagons are passed through the `children` prop */}
+      {props.children}
+    </SharingContext.Provider>
+  );
+};
+```
+
+The `Hexagon` component needs to consume this React context to read its value – here with the `useContext` hook.
+
+```jsx hexagons-30dkf src/Hexagon.js
+const Hexagon = props => {
+  // we get the rendering context by comsuming the React context
+  const renderingContext = React.useContext(SharingContext);
+
+  if (renderingContext !== null) {
+    // hexagon drawing logic
+  }
+};
+```
+
+Now, both our `Canvas` and `Hexagon` components are ready. We can display several random hexagons on the screen!
+
+```jsx hexagons-30dkf src/App.js
+const App = () => (
+  <Canvas>
+    {getRandomHexagons().map(hexagon => (
+      <Hexagon {...hexagon} />
+    ))}
+  </Canvas>
+);
+```
+
+The last thing we need is to animate the hexagons so that they rotate.
+
+## 🎬 Animations
+
+As we saw, the canvas is like a sheet of paper: once we've drawn on it, we can't change it! But we can clear the canvas and re-draw on it. Animating on a canvas is like animating cartoons. It's a lots of drawings quickly shown ones after the others! To make a shape move, we need to split the movement into small steps. Then, we can draw them one by one, while clearing the canvas between them. Those steps are called frames. Browsers come with an API so that we can draw in each frame: `requestAnimationFrame`.
+
+### 🖼 Creating a frame loop
+
+First things first, the canvas should be cleared at each frame. The easiest way to do this is to have an internal state counting the frames. This way, the component re-renders at each frame:
+
+```jsx animated-hexagons-v9j2q src/Canvas.js
+const [frameCount, setFrameCount] = React.useState(0);
+// this effect will increment frameCount by one at the next frame
+// and is called every time frameCount changes, creating a recursive loop
+// this makes the Canvas component re-render at every frame
+React.useEffect(() => {
+  const frameId = requestAnimationFrame(() => {
+    setFrameCount(frameCount + 1);
+  });
+  return () => {
+    cancelAnimationFrame(frameId);
+  };
+}, [frameCount, setFrameCount]);
+
+// here's the clearing at every render — at every frame.
+if (context !== null) {
+  context.clearRect(0, 0, actualWidth, actualHeight);
+}
+```
+
+But... the canvas is now white! It's because we only render the hexagons once – when `<RandomHexagons />` is first rendered. Yet, the canvas is cleared at every frame, erasing the hexagons on the second frame. Children components should also draw in the canvas at every frame. To force them to re-render, one solution is to share the `frameCount` from the canvas with the `Hexagon` component. It's achieved via a React context, like we did with `SharingContext`:
```

Comment:
```suggestion
But... the canvas is now white! It's because the hexagons are only rendered once – when `<RandomHexagons />` is first rendered. Yet, the canvas is cleared at every frame, erasing the hexagons on the second frame. Children components should also draw in the canvas at every frame. To force them to re-render, one solution is to share the `frameCount` from the canvas with the `Hexagon` component. It's achieved via a React context, like we did with `SharingContext`:
```